File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/system/equipment/electrocardiogram/SEElectroCardioGramInterpolationWaveform.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
package mil.tatrc.physiology.datamodel.system.equipment.electrocardiogram;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.ElectroCardioGramInterpolationWaveformData;
import mil.tatrc.physiology.datamodel.bind.EnumHeartRhythm;
import mil.tatrc.physiology.datamodel.properties.SEFunctionElectricPotentialVsTime;
import mil.tatrc.physiology.datamodel.properties.SEScalarTime;
import mil.tatrc.physiology.utilities.Log;

public class SEElectroCardioGramInterpolationWaveform
{
	protected int lead;
  protected EnumHeartRhythm                   rhythm;
  protected SEFunctionElectricPotentialVsTime data;
  protected SEScalarTime                      timeStep;
  
  public SEElectroCardioGramInterpolationWaveform()
  {
    reset();
  }
  
  public void reset()
  {
  	lead = 0;
    rhythm = null;
    data = null;
    timeStep = null;
  }
  
  public boolean load(ElectroCardioGramInterpolationWaveformData in)
  {
    reset();
    setRhythm(in.getRhythm());
    setLead(in.getLead());
    getData().load(in.getData());
    if(in.getTimeStep()!=null)
      getTimeStep().load(in.getTimeStep()); 
    return true;
  }
  
  public ElectroCardioGramInterpolationWaveformData unload()
  {
  	ElectroCardioGramInterpolationWaveformData to =  CDMSerializer.objFactory.createElectroCardioGramInterpolationWaveformData();
    unload(to);
    return to;
  }
  
  protected void unload(ElectroCardioGramInterpolationWaveformData to)
  {
  	to.setLead(lead);
    to.setRhythm(rhythm);
    to.setData(data.unload());
    if(hasTimeStep())
      to.setTimeStep(timeStep.unload());
  }
  
  public int getLead()
  {
    return lead;
  }
  public void setLead(int l)
  {
  	if(l>0 && l<=12)
  		this.lead = l;
  	Log.error("Lead must be between 1 and 12");
  }
  public boolean hasLead()
  {
    return lead != 0;
  }
  public void removeLead()
  {
  	this.lead = 0;
  }
  
  public EnumHeartRhythm getRhythm()
  {
    return rhythm;
  }
  public void setRhythm(EnumHeartRhythm r)
  {
    this.rhythm = r;
  }
  public boolean hasRhythm()
  {
    return rhythm == null ? false : true;
  }
  
  public boolean hasData()
  {
    return data == null ? false : data.isValid();
  }
  public SEFunctionElectricPotentialVsTime getData()
  {
    if (data == null)
      data = new SEFunctionElectricPotentialVsTime();
    return data;
  }

  public boolean hasTimeStep()
  {
    return timeStep == null ? false : timeStep.isValid();
  }
  public SEScalarTime getTimeStep()
  {
    if (timeStep == null)
      timeStep = new SEScalarTime();
    return timeStep;
  }
}

File: projects/biogears/libBiogears/src/cdm/system/equipment/ElectroCardioGram/SEElectroCardioGramInterpolationWaveform.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

#include <biogears/cdm/properties/SEFunctionElectricPotentialVsTime.h>
#include <biogears/cdm/properties/SEScalarElectricPotential.h>
#include <biogears/cdm/system/equipment/ElectroCardioGram/SEElectroCardioGramInterpolationWaveform.h>
#include <biogears/schema/cdm/Physiology.hxx>
#include <biogears/cdm/properties/SEScalarTime.h>

//Private Includes
#include <io/cdm/ElectroCardioGram.h>

namespace biogears {
SEElectroCardioGramInterpolationWaveform::SEElectroCardioGramInterpolationWaveform(Logger* logger)
  : Loggable(logger)
{
  m_TimeStep = nullptr;
  m_Data = nullptr;
  m_Rhythm = (CDM::enumHeartRhythm::value)-1;
  m_LeadNumber = 0;
}
//-----------------------------------------------------------------------------
SEElectroCardioGramInterpolationWaveform::~SEElectroCardioGramInterpolationWaveform()
{
  Clear();
}
//-----------------------------------------------------------------------------
void SEElectroCardioGramInterpolationWaveform::Clear()
{
  m_LeadNumber = 0;
  m_Rhythm = (CDM::enumHeartRhythm::value)-1;
  SAFE_DELETE(m_TimeStep);
  SAFE_DELETE(m_Data);
  m_ActiveIndicies.clear();
}
//-----------------------------------------------------------------------------
bool SEElectroCardioGramInterpolationWaveform::Load(const CDM::ElectroCardioGramInterpolationWaveformData& in)
{
  io::ElectroCardioGram::UnMarshall(in, *this);
  return true;
}
//-----------------------------------------------------------------------------
CDM::ElectroCardioGramInterpolationWaveformData* SEElectroCardioGramInterpolationWaveform::Unload() const
{
  CDM::ElectroCardioGramInterpolationWaveformData* data(new CDM::ElectroCardioGramInterpolationWaveformData());
  Unload(*data);
  return data;
}
//-----------------------------------------------------------------------------
void SEElectroCardioGramInterpolationWaveform::Unload(CDM::ElectroCardioGramInterpolationWaveformData& data) const
{
  io::ElectroCardioGram::Marshall(*this, data);
}
//-----------------------------------------------------------------------------
CDM::ElectroCardioGramWaveformLeadNumberData SEElectroCardioGramInterpolationWaveform::GetLeadNumber() const
{
  return m_LeadNumber;
}
void SEElectroCardioGramInterpolationWaveform::SetLeadNumber(CDM::ElectroCardioGramWaveformLeadNumberData n)
{
  m_LeadNumber = n;
}
bool SEElectroCardioGramInterpolationWaveform::HasLeadNumber() const
{
  return m_LeadNumber == 0 ? false : true;
}
void SEElectroCardioGramInterpolationWaveform::InvalidateLeadNumber()
{
  m_LeadNumber = 0;
}
//-----------------------------------------------------------------------------
CDM::enumHeartRhythm::value SEElectroCardioGramInterpolationWaveform::GetRhythm() const
{
  return m_Rhythm;
}
void SEElectroCardioGramInterpolationWaveform::SetRhythm(CDM::enumHeartRhythm::value rhythm)
{
  m_Rhythm = rhythm;
}
bool SEElectroCardioGramInterpolationWaveform::HasRhythm() const
{
  return m_Rhythm == ((CDM::enumHeartRhythm::value)-1) ? false : true;
}
void SEElectroCardioGramInterpolationWaveform::InvalidateRhythm()
{
  m_Rhythm = (CDM::enumHeartRhythm::value)-1;
}
//-----------------------------------------------------------------------------
bool SEElectroCardioGramInterpolationWaveform::HasData() const
{
  return m_Data == nullptr ? false : m_Data->IsValid();
}
SEFunctionElectricPotentialVsTime& SEElectroCardioGramInterpolationWaveform::GetData()
{
  if (m_Data == nullptr)
    m_Data = new SEFunctionElectricPotentialVsTime();
  return *m_Data;
}
const SEFunctionElectricPotentialVsTime* SEElectroCardioGramInterpolationWaveform::GetData() const
{
  return m_Data;
}
//-----------------------------------------------------------------------------
bool SEElectroCardioGramInterpolationWaveform::HasTimeStep() const
{
  return m_TimeStep == nullptr ? false : m_TimeStep->IsValid();
}
SEScalarTime& SEElectroCardioGramInterpolationWaveform::GetTimeStep()
{
  if (m_TimeStep == nullptr)
    m_TimeStep = new SEScalarTime();
  return *m_TimeStep;
}
double SEElectroCardioGramInterpolationWaveform::GetTimeStep(const TimeUnit& unit) const
{
  if (m_TimeStep == nullptr)
    return SEScalar::dNaN();
  return m_TimeStep->GetValue(unit);
}
//-----------------------------------------------------------------------------
bool SEElectroCardioGramInterpolationWaveform::operator==(SEElectroCardioGramInterpolationWaveform const& rhs) const
{
  if (this == &rhs)
    return true;

  return m_LeadNumber == rhs.m_LeadNumber
    && m_Rhythm == rhs.m_Rhythm
    && ((m_TimeStep && rhs.m_TimeStep) ? m_TimeStep->operator==(*rhs.m_TimeStep) : m_TimeStep == rhs.m_TimeStep)
    && ((m_Data && rhs.m_Data) ? m_Data->operator==(*rhs.m_Data) : m_Data == rhs.m_Data)
    && m_ActiveIndicies == rhs.m_ActiveIndicies;
}
bool SEElectroCardioGramInterpolationWaveform::operator!=(SEElectroCardioGramInterpolationWaveform const& rhs) const
{
  return !(*this == rhs);
}
//-----------------------------------------------------------------------------
}